Add C++ Solution + Editorial to The Great Revegetation (Silver) (#2621)
cpinitiative/usaco-guide: solutions/usaco-920.mdx

id: usaco-920
source: USACO Silver 2019 February
title: The Great Revegetation
author: Sofia Yang, Ryan Chou
---

[Official Analysis (C++)](http://www.usaco.org/current/data/sol_revegetate_silver_feb19.html)

Assume that we have a valid solution composed of $K$ connected components in our graph.

Let's look at one of these connected components, which could look something like this:

![possible connected component](<usaco-920/connected_component.png>)

> However, for all $K$ connected component, the grass types can be flipped and still be a valid solution:

Reviewer: > for all $K$ connected component

component**s**

![inverted connected component](<usaco-920/inverted_connected_component.png>)

There are $2$ possible ways to arrange each connected component. Thus, if there's $k$ connected components, there's $2^k$ ways to arrange the entire graph.

Note how the nodes force their adjacent ones to have a specific coloring, so if any two nodes don't satisfy this condition, it's impossible. Otherwise, the solution will always be $2^k$.
<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	freopen("revegetate.in", "r", stdin);
	freopen("revegetate.out", "w", stdout);
	
	int n, m;
	cin >> n >> m;

	vector<vector<pair<int, bool>>> adj(n);
	
	for (int i = 0; i < m; i++) {
		char type;
		int u, v;
		cin >> type >> u >> v;
		
		adj[--u].push_back({--v, type == 'S'});
		adj[v].push_back({u, type == 'S'});
	}

	int component_num = 0;
	bool impossible = false;
	vector<int> color(n, -1);

	for (int i = 0; i < n; i++) {

> 		if (color[i] == -1) {

Reviewer: comments for this entire part

			component_num++;
			
			queue<pair<int, bool>> todo;
			todo.push({i, true});

> 			while ((int)(todo.size())) {

Reviewer: `!todo.empty()` instead

				pair<int, bool> nxt = todo.front();
				todo.pop();

				color[nxt.first] = nxt.second;

				for (pair<int, bool> u : adj[nxt.first]) {

> 					bool type = (u.second ? nxt.second : !nxt.second);

Reviewer: ```suggestion
					bool type = u.second ? nxt.second : !nxt.second;
```

					
					// haven't visited
					if (color[u.first] == -1) {
						todo.push({u.first, type});
					// creates a contradiction
					} else if (color[u.first] == !type) {
						impossible = true;
						break;
					}
				}
			}
		}
	}

	if (impossible) {
		cout << 0 << endl;
	} else {

> 		/*
> 		 * if we can construct our solution, then there are 2^cc possible rearrangements,
> 		 * since we can always invert the types of grass in each connected component.
> 		 */
> 

Reviewer: ```suggestion
```
alr thoroughly explained in editorial

> 		// 2^cc in binary is 1, followed by cc zeros.

Reviewer: update this comment

id: usaco-920
source: USACO Silver 2019 February
title: The Great Revegetation
author: Sofia Yang, Ryan Chou
---

[Official Analysis (C++)](http://www.usaco.org/current/data/sol_revegetate_silver_feb19.html)

Assume that we have a valid solution composed of $K$ connected components in our graph.

> Now, let's look at one of these, which could look something like this:

Reviewer: ```suggestion
Let's look at one of these connected components, which could look something like this:
```

![possible connected component](<usaco-920/connected_component.png>)

> However, for all $K$ connected component, the grass types can be "inverted":

Author: oops, yeah now that i read this, it sounds really weird

![inverted connected component](<usaco-920/inverted_connected_component.png>)

> So, for each $K$ connected components, there are $2$ possible ways to arrange them, or $2^k$ ways to arrange the entire graph. Since in binary, $2^k$ is $1$ followed by $k$ zeros, we can simply print this instead of manually converting it.

Reviewer: how about
"for each connected component in our graph..."
"thus, if there's k connected components, there's 2^k ways to arrange"

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	freopen("revegetate.in", "r", stdin);
	freopen("revegetate.out", "w", stdout);
	
	int n, m;
	cin >> n >> m;

	vector<vector<pair<int, bool>>> adj(n);
	
	for (int i = 0; i < m; i++) {
		char type;
		int u, v;
		cin >> type >> u >> v;
		
		adj[--u].push_back({--v, type == 'S'});
		adj[v].push_back({u, type == 'S'});
	}

	int cc = 0;
	bool impossible = false;
	vector<int> color(n, -1);

	for (int i = 0; i < n; i++) {
		if (color[i] == -1) {
			cc++;
			
			queue<pair<int, bool>> todo;
			todo.push({i, true});

			while ((int)(todo.size())) {
				pair<int, bool> nxt = todo.front();
				todo.pop();

				color[nxt.first] = nxt.second;

				for (pair<int, bool> u : adj[nxt.first]) {
					bool type = (u.second ? nxt.second : !nxt.second);
					
					// haven't visited
					if (color[u.first] == -1) {
						todo.push({u.first, type});
					// creates a contradiction
					} else if (color[u.first] == !type) {
						impossible = true;
						break;
					}
				}
			}
		}
	}

	if (impossible) {
		cout << 0 << "\n";

> 		return 0;

Reviewer: no need for return 0
